core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -926,6 +937,10 @@ class ReplicaManager(val config: KafkaConfig,
           fetchOnlyFromLeader = fetchOnlyFromLeader,
           minOneMessage = minOneMessage)
 
+
+        // If we are the leader, determine the preferred read-replica
+        val preferredReadReplica = findPreferredReadReplica(tp, clientMetadata, replicaId)
```

Comment:
We are choosing the preferred read replica after reading from the log. I wonder if we should do it first instead. This is related to my other comment, but I think the logic can be something like this:

1. Find all replicas which have data at the fetch offset.
2. If this set is more than just the leader, the select the preferred read replica.
3. If the preferred read replica is not the leader, return empty and skip the read.
4. Otherwise do the read as usual.

Reply:
One more thing, it seems we are selecting the preferred read replica even if we are not the leader? Should we skip this if we are the follower?

Reply:
The logic for checking if we're the leader is inside `findPreferredReadReplica`. I'll update the read logic as suggested